### PR TITLE
[WIP, not for review] TP + DDP support

### PR DESCRIPTION
**Pull request by rohan-varma**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #103867
* #103866
* #103629

Per title

Differential Revision: [D46847183](https://our.internmc.facebook.com/intern/diff/D46847183/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D46847183/)!

### CODE REVIEW

File: test/distributed/tensor/parallel/test_2d_parallel.py [hunk 1]
```diff
@@ -45,12 +47,50 @@ def forward(self, x):
...
+            # m.named_parameters = tp_named_parameters.__get__(m, torch.nn.Module)
+            # # s.make = types.MethodType(times_eight, s, Simple)
+            # m.named_parameters = tp_named_parameters
+        # gen = self._named_members(
+        #     lambda module: module._parameters.items(),
+        #     prefix=prefix, recurse=recurse, remove_duplicate=remove_duplicate)
+        print(m)
+        for mod in m.modules():
+            for param in mod.parameters(recurse=False):
+                if isinstance(param, torch.nn.parameter.UninitializedParameter):
```
Comment (fduwjj) [inline test/distributed/tensor/parallel/test_2d_parallel.py:90]: DTensor is an instance of `UninitializedParameter` which is pretty wired.

File: test/distributed/tensor/parallel/test_2d_parallel.py [hunk 2]
```diff
@@ -127,6 +167,34 @@ def test_2d_fsdp_integration_functionality(self) -> None:
...
+        sd = model_tp.state_dict()
+        load_tp = init_model(use_ddp=True)[0]
+        from torch.distributed._tensor import DTensor
+        for p in load_tp.parameters():
+            if isinstance(p, DTensor):
+                p._local_tensor.data.zero_()
+            else:
+                p.data.zero_()
+
+        load_tp.load_state_dict(sd)
```
Comment (fduwjj) [inline test/distributed/tensor/parallel/test_2d_parallel.py:192]: I appreciate that we keep save/load in mind, but what's more important is the correct of gradient and we are updating model correctly.
